### ISSUE TITLE
Add Broadlink RM Pro Binding

### ISSUE THREAD

**Issue by sidaga** [openhab/openhab-addons#322]
Is it possible to use broadlink rm pro with openhab2? Found a link for sdk for broadlink rm pro.
##

**Comment by kaikreuzer** [comment]
You mean you want to implement a binding for it?

**Comment by sidaga** [comment]
Yes, I think implementing a binding for broadlink rm-pro will enable it to be used with openhab.

Below are some links related to the sdk:

https://drive.google.com/file/d/0B65vYtefY0h2d1NZbEtCUFgwc2hCeHNXMnFmOEdRU1VrcUl3/view?usp=sharing

https://drive.google.com/file/d/0B65vYtefY0h2aE1LdWF5RG9sX00/view?usp=sharing

I don't know where to begin. Last time I checked documentation is in chineese but it was getting translated via google translate. It would be helpful if you can show the path.

Thanks and regards
siddhartha

**Comment by kaikreuzer** [comment]
> I don't know where to begin. Last time I checked documentation is in chineese but it was getting translated via google translate. It would be helpful if you can show the path.
> 
You mean the documentation about openHAB development? It’s the first time I hear that this is only in Chinese :-)

This is where you should find a start: https://github.com/openhab/openhab2/blob/master/docs/sources/development/bindings.md <https://github.com/openhab/openhab2/blob/master/docs/sources/development/bindings.md>

Regards,
Kai

**Comment by sidaga** [comment]
I was talking about the documentation of broadlink rm pro being in
chineese. I apologise if I was not clear about that.

**Comment by kaikreuzer** [comment]
Is this something you will start working on?

**Comment by sidaga** [comment]
Can anyone please guide me on how to proceed with this?

**Comment by sidaga** [comment]
http://milillicuti.com/an-api-to-connect-your-broadlink-rm-pro-irrf-wifi-bridge-to-your-home-automation-system/

**Comment by sidaga** [comment]
http://rm-bridge.fun2code.de/downloads/RMBridge-APIDocumentation.pdf

**Comment by MrBalonio** [comment]
Any progress on this, would like to buy this device

**Comment by bhi-887** [comment]
I am new here - to Open Hab. This post is the only one I found regarding the Broadlink RM2 on OpenHab.
 I have found this regarding the Broadlink RM on desktop outside of OpenHab: 
https://home-assistant.io/components/switch.broadlink/

Their solution is python based. I am not a programmer, so I do not know how much if any of their solution is helpful for use in OpenHab. I am in search of a way to control the Broadlink RM2 pro with a pc rather than a phone for use of by a handicap friend that can use a mouse, but not a touchscreen. 

https://home-assistant.io/getting-started/

**Comment by bhi-887** [comment]
I am new here - to Open Hab. This post is the only one I found regarding the Broadlink RM2 on OpenHab.
 I have found this regarding the Broadlink RM on desktop outside of OpenHab: 
https://home-assistant.io/components/switch.broadlink/

Their solution is python based. I am not a programmer, so I do not know how much if any of their solution is helpful for use in OpenHab. I am in search of a way to control the Broadlink RM2 pro with a pc rather than a phone for use of by a handicap friend that can use a mouse, but not a touchscreen. 

This is the introduction to that program.
https://home-assistant.io/getting-started/

**Comment by mob41** [comment]
@bhi-887 They use the python-broadlink library to have a such binding.

I am implementing a Java version of Broadlink exposed / clean API at https://github.com/mob41/broadlink-java-api

Anyone familiar with Java and python are welcome to contribute to the project.

I am currently stuck at the method to authenticate with broadlink devices.

**Comment by mob41** [comment]
Not to be confused, this does not use the RM Bridge, but directly communicating with the devices. The protocol analysis can be found at https://github.com/mjg59/python-broadlink/blob/master/protocol.md

**Comment by computerlyrik** [comment]
There is also some useful information (code) here: https://github.com/home-assistant/home-assistant/pull/4834/files

As soon as i will receive my A1 Enviromental sensor, i will jump in @mob41's project and include the Sensor-Part into the framework.

In parallel i may help in implementing an broadlink-openhab-addon using this library. (As soon as i did commit my currently working on beta-library for TP-Link HS110 - but that's another story)

**Comment by mob41** [comment]
@computerlyrik Right then, I am going to update/add more documentation about my code (e.g. JavaDoc). I hope you can see whats going wrong in the code. :smile_cat: 

Thanks!

**Comment by mob41** [comment]
Alternatively, I have implemented a binding that uses RMBridge and I am currently using it.

I will soonly push to GitHub and see whether it helps OP

**Comment by dmastag** [comment]
@mob41 did you succeed in doing the binding?
I have the same use-case, any chance you could share how to achieve this using the RMBridge?

**Comment by mob41** [comment]
The problem has been solved last week here mob41/broadlink-java-api#1

I can say that the library is partly successed if I have time to implement all the methods.

**Comment by mob41** [comment]
You can just use an URLConnection to send POST data directly to RMBridge.

As I remember, each command to the RMBridge is clearly stated in the API document provided. And each request is a post request.

StackOverflow have lots of answers about sending a POST request.

**Comment by mob41** [comment]
A quite dirty RMBridge binding for OpenHAB 2... but only ON-OFF switch for Broadlink RM devices. You need to compile it by using the OpenHAB2 Eclipse IDE.

https://github.com/mob41/org.openhab.binding.rmbridge

**Comment by mob41** [comment]
Here's an example how to use RMBridge in Java: 
https://github.com/mob41/org.openhab.binding.rmbridge/blob/master/src/main/java/org/openhab/binding/rmbridge/handler/RMBridgeAPI.java

**Comment by claell** [comment]
There also seems to be another dev who has done a binding, but didn't release right now. Unfortunately there does not seem to be progress right now, but I wanted to mention it: https://community.openhab.org/t/broadlink-binding-for-rmx-a1-spx-and-mp-any-interest/22768

**Comment by jackrobijn** [comment]
Any news on this item.
I have got the A1 unit working in Openhab panel.
![image](https://user-images.githubusercontent.com/15873884/31114547-f12f3018-a81e-11e7-8216-0fa05da48fe6.png)

But strugling to get the rm2 unit working.
Used the binding from Cato and added the A1 RM2 and RM3 device.
Managed to read the commands with RM bridge for RM2, but unable to send commands to RM2
Temperature reading from RM2 is working, any help is much appriciated

**Comment by mob41** [comment]
Which binding are you using? What I currently know is there aren't a proper Broadlink OpenHUB binding available in the internet. Can you give me a link?

**Comment by jackrobijn** [comment]
@mob41,
i have used the broadlink binding written by Cato Sognen, i you google on :Broadlink binding for RMx, A1, SPx and MP. Any interest? you will find all info from Cato: If you start reading as from 2 july, you will find all the info

**Comment by ARVINTHMOORTHY** [comment]
Sir,
I am planning to make automation in my home. Am new to this field. I have searched lot of device in web. And I found broadlink rm pro+ gateway in Indian web. Which has IR , RF & WIFI. Am planning to use more RF sensors and relay switches in my home. My query is 
1. multiple rf (433mah) devices will operated each other while operating at a time.
2. Does all rf devices available in local market have different rf codes to configure with the gateway.
3. Any interference will occur?
4. If automation scene is set, will the gateway operate the preset rf devices. 
5. Any interference will occur between IR & RF & WIFI devices.

Kindly give ur suggestions plz
ARVINTH
arvibluedragon@gmail.com

**Comment by claell** [comment]
@ARVINTHMOORTHY This is not the right place for this questions. However, maybe someone in the OpenHAB Forum can help you. I also linked a thread for the binding in the comments above.
Also as a hint: If your sentence is a question, use a question mark to indicate it.

**Comment by claell** [comment]
@davidgraeff There is still an ongoing effort for this binding in the thread I linked a few posts above. Now (since a few months) @themillhousegroup has taken over the development from the original author in that thread and also has a GitHub repo for this. Hopefully it is only a matter of time until an official PR is created for this, so this is not out of scope and thus should not be closed, correct?

**Comment by davidgraeff** [comment]
@themillhousegroup can you create your own pull request please?
We use pull requests for code feedback and merge preparation. If the idea is to just indicate that this feature is being worked on, an Issue is the better place :)

**Comment by J-N-K** [comment]
Any progress here?

**Comment by themillhousegroup** [comment]
@J-N-K the Broadlink binding is still a work-in-progress but as of last week I successfully have a 2.5 version built and working under OpenHAB 2.5. I have a few things to tidy up but have every intention of issuing a PR to get this into the official repo very shortly.

If you can't wait, grab the JAR from here: 
https://github.com/themillhousegroup/openhab2-addons/releases/tag/BROADLINK_2.5.BETA_01
